docs: updated documentation for MVC component. (#245)
mautic/developer-documentation-new: docs/plugins/mvc.rst

   * - redirect
     - OPTIONAL
     - string
     - The URL to force a page redirect instead of injecting AJAX content.
   * - target
     - OPTIONAL
     - string
     - jQuery selector to inject the content into. Defaults to app’s main content selector.
   * - replaceContent
     - OPTIONAL
     - string
     - If set to `'true'`, Mautic replaces the target selector with AJAX content.

delegateRedirect($url) 
^^^^^^^^^^^^^^^^^^^^^^

Delegates the appropriate response for redirects.

- **If AJAX request**: returns a json response with ``{redirect: $url}``.  
- **If http request**: performs a standard redirect header.

postActionRedirect($args)
^^^^^^^^^^^^^^^^^^^^^^^^^

Similar to ``delegateView()``, but used after an action like saving a Form. Accepts the same ``$args`` as ``delegateView()``, plus:

.. list-table:: Additional Parameters for ``postActionRedirect()``
   :widths: 20 10 10 60
   :header-rows: 1

   * - Key
     - Required
     - Type
     - Description
   * - ``returnUrl``
     - OPTIONAL
     - string
     - URL to redirect to. Defaults to ``/s/dashboard``. Auto-populates ``passthroughVars[route]`` if not set.
   * - flashes
     - OPTIONAL
     - array

>      - Array of flash messages to display after redirecting. See :doc:`Flash Messages <rest_api/assets>` for more information.

Reviewer: Is the link to Flash Messages correct?

https://devdocs.mautic.org/en/5.x/rest_api/assets.html

Reviewer 2: Nope. Perhaps we could remove the link?

   * - ``forwardController``
     - OPTIONAL
     - bool
     - If true (default), forwards to a controller method (``BundleName:ControllerName:method``). Set to ``false`` to load a view template (``BundleName:ViewName:template.html.php``) directly.

2. FormController (``Mautic\CoreBundle\Controller\FormController``)
-------------------------------------------------------------------

> This controller extends ``CommonController`` and provides helper methods for managing :doc:`Forms <components/forms>`.

Reviewer: Is the link to the Forms here correct?

https://devdocs.mautic.org/en/5.x/components/forms.html

Reviewer 2: Probably not. I'd remove the link

                        [
                            'returnUrl'       => $this->generateUrl('plugin_helloworld_world'),
                            'contentTemplate' => 'MauticPlugin\HelloWorldBundle\Controller\DefaultController:worldAction',
                            'flashes'         => $flashes
                        ]
                    );
                } // Otherwise show the form again with validation error messages
            }

            // Display the form
            return $this->delegateView(
                [
                    'viewParameters'  => [
                        'form' => $form->createView()
                    ),
                    'contentTemplate' => '@HelloWorld/Contact/form.html.twig',
                    'passthroughVars' => [
                        'activeLink' => 'plugin_helloworld_contact',
                        'route'      => $this->generateUrl('plugin_helloworld_contact')
                    ]
                ]
            );
        }
    }

3. AjaxController (``Mautic\CoreBundle\Controller\AjaxController``)
-------------------------------------------------------------------

> This controller also extends ``CommonController`` and is a companion to some of the built-in JavaScript helpers. See *JavaScript methods* for more information.

Reviewer: This link is empty both here and in the legacy docs. Where this should goes to?

Reviewer 2: I'd remove the link

            );

            $mailer->message->setSubject($data['subject']);

            $mailer->message->setBody($data['message']);

            $mailer->send();
        }
    }

Base Model Classes
------------------

You can extend either of the following base classes to make use of Mautic's helper methods:

1. AbstractCommonModel (``\Mautic\CoreBundle\Model\AbstractCommonModel``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This base class offers access to services commonly used in models:

.. list-table::
    :header-rows: 1

    * - Property
      - Service
      - Description
    * - ``$this->em``
      - Entity manager
      - Handles database interactions via Doctrine.
    * - ``$this->security``
      - Security service
      - Provides access to the current user and permission checks.
    * - ``$this->dispatcher``
      - Event dispatcher
      - Dispatches and listens for Mautic events.
    * - ``$this->translator``
      - Translator service
      - Handles language translations.

2. FormModel (``\Mautic\CoreBundle\Model\FormModel``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This extends ``AbstractCommonModel`` and includes helper methods for working with entities and repositories. For more information, see the :doc:`Database <components/entities>` section.

Reviewer: Is the link to Database here correct?

https://devdocs.mautic.org/en/5.x/plugins/data.html

The section's title should be "Entities and schema" instead of "Database" here.

.. code-block:: php

    // Dynamically insert script into head
    echo $view['assets']->includeScript('plugins/HelloWorldBundle/assets/helloworld.js');

Inserting stylesheets
~~~~~~~~~~~~~~~~~~~~~

Use ``includeStylesheet()`` to dynamically include a CSS file into the head.

.. code-block:: php

    // Dynamically insert stylesheet into head
    echo $view['assets']->includeStylesheet('plugins/HelloWorldBundle/assets/helloworld.css');

These methods ensure that your assets are properly handled regardless of Mautic’s installation location or environment. They also support dynamic inclusion for content loaded via AJAX.

Router helper
-------------

The ``router`` helper, accessed via ``$view['router']``, is used to generate URLs to named routes within views.

.. code-block:: php

    <a href="<?php echo $view['router']->generate(
        'plugin_helloworld_world',
        array('world' => 'mars')
    ); ?>" data-toggle="ajax">Mars</a>

This generates a link to the route ``plugin_helloworld_world`` with the dynamic parameter ``world`` set to ``mars``.

> For more details on defining and using routes, see :doc:`Router </plugins/config>`.

Reviewer: Is the link to Router here is correct?

https://devdocs.mautic.org/en/5.x/plugins/config.html

Reviewer 2: Nope, not correct

Translation helper
------------------

The ``translator`` helper, accessed via ``$view['translator']``, is used to translate strings within views using Mautic's translation system.

.. code-block:: php

    <h1>
        <?php echo $view['translator']->trans(
            'plugin.helloworld.worlds',
            array('%world%' => 'Mars')
        ); ?>
    </h1>

This example replaces the ``%world%`` placeholder with ``Mars``, and output the translated string.

> For more on how to handle translations, see :doc:`Translator </developer/translations>`.

Reviewer: Is this the below is the correct link for Translator here?

https://devdocs.mautic.org/en/5.x/components/translator.html

Reviewer 2: I think so

    // Format using short date-time format
    $shortDateTime = $view['date']->toShort($datetime);

    // Format using date-only format
    $date = $view['date']->toDate($datetime);

    // Format using time-only format
    $time = $view['date']->toTime($datetime);

    // Combine date-only and time-only formats
    $datetime = $view['date']->toFullConcat($datetime);

    // Format as relative time: 'Yesterday, 8:02 pm' or 'x days ago'
    $text = $view['date']->toText($datetime);

    // Format a date string in a different timezone
    $fullDateTime = $view['date']->toFull($datetime, 'Y-m-d H:i:s', 'UTC');

The first argument to each method can be a ``\DateTime`` object or a string formatted as ``Y-m-d H:i:s``. If the date is not already in local time, pass the expected format as the second argument and the timezone as the third.

Form helper
-----------

The ``form`` helper, accessed via ``$view['form']``, is used to render form objects passed from the controller.

.. code-block:: php

    <?php echo $view['form']->form($form); ?>

This helper outputs the full HTML Form using the Form object (typically a Symfony Form) passed to the view.

> For detailed usage, see :doc:`Forms <components/forms.`.

Reviewer: Is the link to Forms here correct?

https://devdocs.mautic.org/en/5.x/components/forms.html

Reviewer 2: Probably not. I'd remove the link

MVC
###

> Mautic uses a **Model-View-Controller (MVC)** structure to manage how users interact with the frontend (**views**) and how the backend handles those interactions (**controllers and models**). Additionally, **Entity** and **Repository** classes manage interactions with the database.

Reviewer 2: As the Model layer handles data manipulation then repositories and entities also belong to the Model layer. Unsure how important it is to mention. But other classes still fit under one of the 3 layers.

Reviewer: So, I think we can safely remove the second sentence?

Reviewer 2: Agree

In Symfony, and thus Mautic  the **controller** is the central part of the MVC structure. The route determines which controller method executes when a user makes a request. The controller then interacts with the **model** to retrieve or manipulate data, and finally renders a **view** to display the results to the user.

Controllers
===========
Matching Routes to controller methods
-------------------------------------

The  :ref:`route defined in the config</plugins/config>` determines which controller method is called. Take this example:

.. code-block:: php

    'plugin_helloworld_admin' => array(
        'path'       => '/hello/admin',
        'controller' => 'HelloWorldBundle:Default:admin'

>     ),

Reviewer 2: More modern syntax, changing the route to Symfony 5+
```suggestion
    'plugin_helloworld_admin' => [
        'path'       => '/hello/admin',
        'controller' => 'MauticPlugin\HelloWorldBundle\Controller\DefaultController:adminAction'
    ],
```
This new path is longer but anyone now understands what the class and method is. Perhaps we can remove the lines bellow that were explaining the older version.

Reviewer: You mean these lines?

```rst
The system identifies the controller as``HelloWorldBundle:Default:admin``. Broken down, that translates to:

- ``HelloWorldBundle`` → ``\MauticPlugin\HelloWorldBundle\Controller``
- ``Default`` → ``DefaultController``
- ``admin`` → ``adminAction()``

.. note::

   Controller notation follows the format ``BundleName:ControllerName:controllerMethod.``

   To use a controller within a subfolder of ``Controller``, use this format ``BundleName:Subdirectory\\ControllerName:controllerMethod``
```

Reviewer 2: Yes

The system identifies the controller as``HelloWorldBundle:Default:admin``. Broken down, that translates to:

- ``HelloWorldBundle`` → ``\MauticPlugin\HelloWorldBundle\Controller``
- ``Default`` → ``DefaultController``
- ``admin`` → ``adminAction()``

.. note::

   Controller notation follows the format ``BundleName:ControllerName:controllerMethod.``

   To use a controller within a subfolder of ``Controller``, use this format ``BundleName:Subdirectory\\ControllerName:controllerMethod``

Thus, when a browser calls up ``/hello/admin``, ``\MauticPlugin\HelloWorldBundle\Controller\DefaultController::adminAction()`` will be called.

Route placeholders
------------------

Symfony automatically passes route placeholders into the controller’s method as arguments. The method’s parameters must match the placeholder names.

Example:

.. code-block:: php

    'plugin_helloworld_world' => array(
        'path'       => '/hello/{world}',
        'controller' => 'HelloWorldBundle:Default:world',
        'defaults'    => array(
            'world' => 'earth'
        ),
        'requirements' => array(
            'world' => 'earth|mars'
        )

>     ),

Reviewer 2: ```suggestion
    'plugin_helloworld_world' => [
        'path'       => '/hello/{world}',
        'controller' => 'MauticPlugin\HelloWorldBundle\Controller\DefaultController::worldAction',
        'defaults'    => [
            'world' => 'earth'
        ],
        'requirements' => [
            'world' => 'earth|mars'
        ]
    ],
```

The matching method:

.. code-block:: php

>     public function worldAction($world = 'earth')

Reviewer 2: ```suggestion
    public function worldAction(string $world = 'earth')
```

.. note::

   Since the route defines a default for ``world``, the controller method must also reflect this default.

If the route looked like this instead:

.. code-block:: php

    'plugin_helloworld_world' => array(
        'path'       => '/hello/{world}',
        'controller' => 'HelloWorldBundle:Default:world',
        'requirements' => array(
            'world' => 'earth|mars'
        )

>     ),

Reviewer 2: ```suggestion
    'plugin_helloworld_world' => [
        'path'       => '/hello/{world}',
        'controller' => 'MauticPlugin\HelloWorldBundle\Controller\DefaultController::worldAction',
        'requirements' => [
            'world' => 'earth|mars'
        ]
    ],
```

Then the method must be:

.. code-block:: php

>     public function worldAction($world)

Reviewer 2: ```suggestion
    public function worldAction(string $world)
```

Extending Mautic’s controllers
------------------------------

Mautic has several controllers that provide some helper functions.

1. CommonController (``Mautic\CoreBundle\Controller\CommonController``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Controllers extending this makes ``MauticFactory`` available via ``$this->factory`` and ``Request`` via ``$this->request``.

Reviewer 2: There is no MauticFactory anymore. Let's remove this line.
```suggestion
```

    {
        /**
         * Display the world view
         *
         * @param string $world
         *
         * @return JsonResponse|\Symfony\Component\HttpFoundation\Response
         */
        public function worldAction($world = 'earth')
        {
            /** @var \MauticPlugin\HelloWorldBundle\Model\WorldModel $model */
            $model = $this->getModel('helloworld.world');

            // Retrieve details about the world
            $worldDetails = $model->getWorldDetails($world);

            return $this->delegateView(
                array(
                    'viewParameters'  => array(
                        'world'   => $world,
                        'details' => $worldDetails
                    ),
                    'contentTemplate' => 'HelloWorldBundle:World:details.html.php',
                    'passthroughVars' => array(
                        'activeLink'    => 'plugin_helloworld_world',
                        'route'         => $this->generateUrl('plugin_helloworld_world', array('world' => $world)),
                        'mauticContent' => 'helloWorldDetails'
                    )
                )
            );

>         }

Reviewer 2: ```suggestion
        public function worldAction(string $world = 'earth', WorldModel $model): Response
        {
            // Retrieve details about the world
            $worldDetails = $model->getWorldDetails($world);

            return $this->delegateView(
                [
                    'viewParameters'  => [
                        'world'   => $world,
                        'details' => $worldDetails
                    ],
                    'contentTemplate' => 'HelloWorldBundle:World:details.html.php',
                    'passthroughVars' => [
                        'activeLink'    => 'plugin_helloworld_world',
                        'route'         => $this->generateUrl('plugin_helloworld_world', ['world' => $world]),
                        'mauticContent' => 'helloWorldDetails'
                    ]
                ]
            );
        }
```

                    }
                }

                if ($cancelled || $valid) {
                    // Redirect to /hello/world

                    return $this->postActionRedirect(
                        array(
                            'returnUrl'       => $this->generateUrl('plugin_helloworld_world'),
                            'contentTemplate' => 'HelloWorldBundle:Default:world',
                            'flashes'         => $flashes
                        )
                    );
                } // Otherwise show the form again with validation error messages
            }

            // Display the form
            return $this->delegateView(
                array(
                    'viewParameters'  => array(
                        'form' => $form->createView()
                    ),
                    'contentTemplate' => 'HelloWorldBundle:Contact:form.html.php',
                    'passthroughVars' => array(
                        'activeLink' => 'plugin_helloworld_contact',
                        'route'      => $this->generateUrl('plugin_helloworld_contact')
                    )
                )
            );
        }

>     }

Reviewer 2: ```suggestion
        public function contactAction(ContactModel $model): Response
        {
            // Create the form object
            $form = $this->formFactory->create(ContactFormType::class);

            // Handle form submission if POST        
            if ($this->request->getMethod() == 'POST') {
                $flashes = [];

                // isFormCancelled() checks if the cancel button was clicked
                if ($cancelled = $this->isFormCancelled($form)) {

                    // isFormValid() will bind the request to the form object and validate the data
                    if ($valid = $this->isFormValid($form)) {

                        // Send the email
                        $model->sendContactEmail($form->getData());

                        // Set success flash message
                        $flashes[] = [
                            'type'    => 'notice',
                            'msg'     => 'plugin.helloworld.notice.thank_you',
                            'msgVars' => [
                                '%name%' => $form['name']->getData()
                            ]
                        ];
                    }
                }

                if ($cancelled || $valid) {
                    // Redirect to /hello/world

                    return $this->postActionRedirect(
                        [
                            'returnUrl'       => $this->generateUrl('plugin_helloworld_world'),
                            'contentTemplate' => 'MauticPlugin\HelloWorldBundle\Controller\DefaultController:worldAction',
                            'flashes'         => $flashes
                        ]
                    );
                } // Otherwise show the form again with validation error messages
            }

            // Display the form
            return $this->delegateView(
                [
                    'viewParameters'  => [
                        'form' => $form->createView()
                    ),
                    'contentTemplate' => '@HelloWorld/Contact/form.html.twig',
                    'passthroughVars' => [
                        'activeLink' => 'plugin_helloworld_contact',
                        'route'      => $this->generateUrl('plugin_helloworld_contact')
                    ]
                ]
            );
        }
    }
```

Models
======

Models retrieve and process data between controllers and views. While not required in Plugins, Mautic provides convenient ways to access model objects and use commonly needed methods if you choose to use them.

Model example
-------------

.. code-block:: php

    <?php
    // plugins/HelloWorldBundle/Model/ContactModel.php

    namespace MauticPlugin\HelloWorldBundle\Model;

    use Mautic\CoreBundle\Model\CommonModel;

    class ContactModel extends CommonModel
    {
        /**
         * Send contact email
         * 
         * @param array $data
         */
        public function sendContactEmail($data)
        {
            // Get mailer helper - pass the mautic.helper.mailer service as a dependency
            $mailer = $this->mailer;

            $mailer->message->addTo(
                $this->factory->getParameter('mailer_from_email')
            );

            $this->message->setFrom(
                array($data['email'] => $data['name'])
            );

            $mailer->message->setSubject($data['subject']);

            $mailer->message->setBody($data['message']);

            $mailer->send();

>         }

Reviewer 2: ```suggestion
    final class ContactModel extends CommonModel
    {
        /**
         * Send contact email
         */
        public function sendContactEmail(array $data): void
        {
            $mailer->message->addTo(
                $this->coreParametersHelper->get('mailer_from_email')
            );

            $this->message->setFrom(
                array($data['email'] => $data['name'])
            );

            $mailer->message->setSubject($data['subject']);

            $mailer->message->setBody($data['message']);

            $mailer->send();
        }
```

    }

Base Model Classes
------------------

You can extend either of the following base classes to make use of Mautic's helper methods:

1. AbstractCommonModel (``\Mautic\CoreBundle\Model\AbstractCommonModel``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This base class offers access to services commonly used in models:

.. list-table::
    :header-rows: 1

    * - Property
      - Service
      - Description
    * - ``$this->factory``
      - Factory service

>       - Provides access to other Mautic services. *Deprecated as of Mautic 2.0*; use dependency injection instead.

Reviewer 2: ```suggestion
```
No longer available

    * - ``$this->em``
      - Entity manager
      - Handles database interactions via Doctrine.
    * - ``$this->security``
      - Security service
      - Provides access to the current user and permission checks.
    * - ``$this->dispatcher``
      - Event dispatcher
      - Dispatches and listens for Mautic events.
    * - ``$this->translator``
      - Translator service
      - Handles language translations.

2. FormModel (``\Mautic\CoreBundle\Model\FormModel``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This extends ``AbstractCommonModel`` and includes helper methods for working with entities and repositories. For more information, see the :doc:`Database <components/entities>` section.

Getting Model Objects
---------------------

To retrieve a model object in a controller:

.. code-block:: php

    <?php

    /** @var \Mautic\LeadBundle\Model\LeadModel $leadModel */
    $leadModel = $this->getModel('lead'); // Shortcut for lead.lead

    /** @var \Mautic\LeadBundle\Model\ListModel $leadListModel */
    $leadListModel = $this->getModel('lead.list');

    /** @var \MauticPlugin\HelloWorldBundle\Model\ContactModel $contactModel */

>     $contactModel = $this->getModel('helloworld.contact');

Reviewer 2: ```suggestion
To retrieve a model object in a controller use Symfony's autowiring (https://symfony.com/doc/current/controller.html#controller-accessing-services)
```
Getting dependencies as strings is no longer recommended. I think we can just link to the symfony docs.

Reviewer: We have an autowiring link as below:

https://symfony.com/doc/5.x/service_container/autowiring.html

Do you mean **Symfony's fetching services** in the suggestion?

If using a model inside another service or model, inject the model service as a dependency instead of using the helper method.

Views
=====

Views in Mautic take data passed from the controller and display it to the user. You can render templates from within controllers or other templates.

The controller uses the ``delegateView()`` method to render views, which relies on the ``contentTemplate`` to determine which view to render.

View notation follows this format:

.. code-block:: none

>     BundleName:ViewName:template.html.php

Reviewer 2: ```suggestion
    @BundleName/ViewName/template.html.twig
```

> For example, ``HelloWorldBundle:Contact:form.html.php`` points to the file ``/path/to/mautic/plugins/HelloWorldBundle/Views/Contact/form.html.php``

Reviewer 2: ```suggestion
For example, ``@HelloWorld/Contact/form.html.twig`` points to the file ``/path/to/mautic/plugins/HelloWorldBundle/Resources/views/Contact/form.html.twig``
```

> To use views inside subfolders under ``Views``, use backslash notation:

Reviewer 2: ```suggestion
To use views inside subfolders under ``Resources/views``:
```

.. code-block:: none

>     BundleName:ViewName\Subfolder:template.html.php

Reviewer 2: ```suggestion
    @BundleName/ViewName/Subfolder/template.html.twig
```

View parameters
---------------

The array passed as ``viewParameters`` in the controller’s ``delegateView()`` method becomes available as variables in the view.

For example, if the controller passes:

.. code-block:: php

    'viewParameters' => array(
        'world' => 'mars'

>     ),

Reviewer 2: ```suggestion
    'viewParameters' => [
        'world' => 'mars'
    ],
```

Then the variable ``$world`` becomes available in the template with the value ``mars``.

Some variables are always available and shouldn't be overridden:

- ``$view``: contains helper objects for extending or rendering templates.
- ``$app``: provides access to request and session objects via ``$app->getRequest()`` and ``$app->getSession()``.

Extending Views

> ---------------

Reviewer 2: The view system changed from PHP templates to Twig in Mautic 5. The following is completely wrong. Can we remove it from this PR? I think we should start over with view docs or just link to the twig docs instead until someone gets to it?

Reviewer: Is this the correct Twig docs? https://twig.symfony.com/doc/3.x/tags/extends.html

Maybe for now, we can do as so?

```suggestion
Extending Views
---------------

Please read the Twig docs to learn about extending views.
```

Reviewer 2: Yes, that's it
